valhalla/thor/dijkstras.h

```diff
@@ -194,29 +236,187 @@ class Dijkstras {
-                     const baldr::TimeInfo& time_info);
+  template <const ExpansionType expansion_direction,
+            // As multimodal isn't ported yet, but we're using the 3-value enum, check that
+            // the caller is using a supported expansion mode (this is a compile time check to
+            // catch this mistake early).  Once multimodal is supported, this can be removed.
+            class = typename std::enable_if<expansion_direction == ExpansionType::forward ||
+                                            expansion_direction == ExpansionType::reverse>::type>
+  void ExpandTemplate(baldr::GraphReader& graphreader,
+                      const baldr::GraphId& node,
+                      const sif::EdgeLabel& pred,
+                      const uint32_t pred_idx,
+                      const baldr::DirectedEdge* opp_pred_edge,
+                      const bool from_transition,
+                      const baldr::TimeInfo& time_info) {
+
+    // Get the tile and the node info. Skip if tile is null (can happen
+    // with regional data sets) or if no access at the node.
+    graph_tile_ptr tile = graphreader.GetGraphTile(node);
+    if (tile == nullptr) {
+      return;
+    }
+
+    // Get the nodeinfo
+    const baldr::NodeInfo* nodeinfo = tile->node(node);
+
+    // We dont need to do transitions again we just need to queue the edges that leave them
+    if (!from_transition) {
+      // Let implementing class know we are expanding from here
+      sif::EdgeLabel* prev_pred =
+          pred.predecessor() == baldr::kInvalidLabel ? nullptr : &bdedgelabels_[pred.predecessor()];
+      ExpandingNode(graphreader, tile, nodeinfo, pred, prev_pred);
+    }
+
+    // Bail if we cant expand from here
+    if (!costing_->Allowed(nodeinfo)) {
+      return;
+    }
+
+    // Update the time information
+    auto offset_time =
+        from_transition
+            ? time_info
+            : (expansion_direction == ExpansionType::forward
+                   ? time_info.forward(pred.cost().secs, static_cast<int>(nodeinfo->timezone()))
+                   : time_info.reverse(pred.cost().secs, static_cast<int>(nodeinfo->timezone())));
+
+    // Expand from end node in forward direction.
+    baldr::GraphId edgeid = {node.tileid(), node.level(), nodeinfo->edge_index()};
+    EdgeStatusInfo* es = edgestatus_.GetPtr(edgeid, tile, pred.path_id());
+    const baldr::DirectedEdge* directededge = tile->directededge(edgeid);
+    for (uint32_t i = 0; i < nodeinfo->edge_count(); ++i, ++directededge, ++edgeid, ++es) {
+      // Skip this edge if permanently labeled (best path already found to this
+      // directed edge). skip shortcuts or if no access is allowed to this edge
+      // (based on the costing method) or if a complex restriction exists for
+      // this path.
+      if (expansion_direction == ExpansionType::forward) {
+        if (directededge->is_shortcut() || es->set() == EdgeSet::kPermanent ||
```

Comment:
@dnesbitt61 In the original code, you had authored this `if` statement and the one below it (line 299) with a different order of operations.  I've normalized them for readability, but was there a reason for the different ordering of conditions?

Reply:
I benchmarked and tested the consolidated version of this - I see no difference in performance or behaviour.